flowmachine/flowmachine/core/subscriber_subset.py

```diff
@@ -0,0 +1,186 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from abc import abstractmethod
+from sqlalchemy.sql import ClauseElement, select, text, column
+from .query import Query
+
+
+class SubscriberSubsetBase(Query):
+    """
+    Base class for the different types of subscriber subsets.
+    """
+
+    @property
+    @abstractmethod
+    def is_proper_subset(self):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'is_proper_subset'"
+        )
+
+    @abstractmethod
+    def apply_subset(self, sql, *, subscriber_identifier):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'apply_subset'"
+        )
+
+    def _get_query_attrs_for_dependency_graph(self, analyse=False):
+        # This is a stub implementation of this internal method.
+        # It is needed because SubscriberSubsetBase currently
+        # inherits from flowmachine.Query, so we implement just
+        # enough to ensure Query.dependency_graph() doesn't break.
+        attrs = {}
+        attrs["name"] = self.__class__.__name__
+        attrs["stored"] = "N/A"
+        attrs["cost"] = "N/A"
+        attrs["runtime"] = "N/A"
+        return attrs
+
+
+class AllSubscribers(SubscriberSubsetBase):
+    """
+    Represents the subset of all subscribers (i.e., no subsetting at all).
+    """
+
+    is_proper_subset = False
+
+    def _make_query(self):
```

Comment:
So what happens if another Query tries to make use of one of these subscriber objects? The string from `get_query` will not be valid sql?

Reply:
Yes. This is clearly not great but it should only be transitional until we can avoid the subsets being inherited from `Query`, and no other query should need to make use of this string other than in generating the md5 hash. Now that you mention it I do wonder, though, how this will affect the caching and whether we need to include the arguments in the string returned from here...